pages/design_documentation/cleaner.md

```diff
@@ -5,12 +5,138 @@ toc: false
+**ALRU** is selected by default.
+
+The cleaning is performed by periodically invoking *ocf_cleaner_run()*. Once sanity 
+checks are performed, a cleaning policy specific *perform_cleaning()* operation 
+is called. After determining what data is to be cleaned, the policy then invokes 
+the cleaner function - *ocf_cleaner_fire()*. This function performs the actual 
+flushing, moving the dirty data to a core device.
+
 
 ## Cleaning policies
+The algorithm used for making a decision what and when to clean is determined 
+by the cleaning policy, which is selected during cache startup and may be changed
+during runtime. 
+
+Currently, OpenCAS supports the following cleaning policies:
+
 ### ALRU - Approximately Least Recently Used
+Approximately Least Recently Used is the default cleaning policy if not specified
+otherwise. It is based on a classic least-recently-used list (LRU) approach.
+
+
+#### Configuration
+ALRU policy uses several config parameters to tune the behavior of the cleaner.
+
+* **Wake Up**- cleaning thread idle wake up time. This happens when there is no data
+to clean or when the thread didn't start due to activity time limit (default: 20s)
+* **Staleness Time** - the time a cache line resides in the LRU list before cleaning
+attempt (default: 120s), 
+* **Flush Max Buffers** - maximum amount of dirty cache lines to flush in one cleaning 
+operation (default: 100 lines)
+* **Activity Threshold** - cache I/O activity cool-down time period. Cleaning will
+not be commenced before this time period has expired (default: 10000 ms) 
+
+
+#### Operation
+The LRU list grows as the dirty data are being produced - when a cache line is 
+hot, i.e. has been just written by the **write-back engine**, it is added to 
+the beginning of the LRU list, tagged with current timestamp. 
+
+If a cache line already exists on the list, it is removed from the list only to 
+be inserted again at the beginning, now with a current timestamp.
+
+After **Wake Up** time has passed, background cleaning thread is woken up and
+tries to perform cleaning. However, cleaning will only happen when cache I/O 
+**activity threshold** time has expired.
+
+If all the conditions are met, a flush list of up to **Flush Max Buffers** 
+is built by removing the cache lines from the back of the LRU list if their 
+**staleness time** has expired.
+
+This list is then provided to the OCF cleaner by issuing *ocf_cleaner_fire()*. 
+OCF cleaner then issues an asynchrounous flushing request, eventually writing 
+the dirty data to the core device.
+
+This process is repeated until there is no further data to clean. The 
+background task is then put to sleep for **Wake Up** time period.
+
+
 ### ACP - Aggressive Cleaning Policy
-### NOP - No Operation (effectively disables background cleaning)
+Aggressive Cleaning Policy takes a rather different approach than ALRU. It organizes
+the data into 100MiB chunks, maintaining a list of them per core device. The 
```

Comment:
"It organizes the data into 100MiB chunks" - in what way?
I think we want to emphasize that the core device is split into those chunks, and the percentage of dirty cache lines is counted for each chunk separately, and based on this statistic chunks are moved between the buckets.
I know this information is revealed later, but this information is key to understand what ACP really is. Plus when we put it this way the information "maintaining a list of them per core device" becomes unnecessary, as at that point it's obvious that those chunks are fragments of core device address space.